django/db/models/fields/related_descriptors.py

```diff
@@ -164,23 +166,19 @@ def get_prefetch_querysets(self, instances, querysets=None):
         rel_obj_attr = self.field.get_foreign_related_value
         instance_attr = self.field.get_local_related_value
         instances_dict = {instance_attr(inst): inst for inst in instances}
-        related_field = self.field.foreign_related_fields[0]
+        related_fields = self.field.foreign_related_fields
         remote_field = self.field.remote_field
-
-        # FIXME: This will need to be revisited when we introduce support for
-        # composite fields. In the meantime we take this practical approach to
-        # solve a regression on 1.6 when the reverse manager is hidden
-        # (related_name ends with a '+'). Refs #21410.
-        # The check for len(...) == 1 is a special case that allows the query
-        # to be join-less and smaller. Refs #21760.
-        if remote_field.hidden or len(self.field.foreign_related_fields) == 1:
-            query = {
-                "%s__in"
-                % related_field.name: {instance_attr(inst)[0] for inst in instances}
-            }
-        else:
-            query = {"%s__in" % self.field.related_query_name(): instances}
-        queryset = queryset.filter(**query)
+        queryset = queryset.filter(
+            TupleIn(
+                ColPairs(
+                    queryset.model._meta.db_table,
+                    related_fields,
+                    related_fields,
+                    self.field,
+                ),
+                list(instances_dict),
+            )
+        )
```

Comment:
This could also be turned into the following if we want to keep special casing 1 or many fields but it didn't seem worth it

```suggestion
        if len(related_fields) == 1:
            query = Q.create(
                [
                    (
                        "%s__in" % related_fields[0].name,
                        {instance_attr(inst)[0] for inst in instances},
                    )
                ]
            )
        else:
            query = TupleIn(
                ColPairs(
                    queryset.model._meta.db_table,
                    related_fields,
                    related_fields,
                    self.field,
                ),
                list(instances_dict),
            )
        queryset = queryset.filter(query)
```